Add infinite codegen stress feature to crossgen2 and fix memory usage issues found while doing so

davidwrighton opened the pull request:
- Add Inifinite codegen stress mode which runs the compilation portion of crossgen2 in an infinite loop
- Fix a couple of memory leaks around PInvokeTargetNativeMethod and UnboxingMethodDesc
  - These were being stored into long-lived data structures, and I was able to make at least UnboxingMethodDesc have a feature which prevents them from being stored for a long period of time
- Address a number of scenarios where we allocate tremendous amounts of useless garbage
  - Our pattern for using ConcurrentDictionary was allocating a delegate on every lookup
  - Tweak ModuleTokenResolver to not need to allocate new copies of the `TokenResolverProvider` and the various `ImmutableArray` objects in the signature decoders as we don't actually need the return value from those apis.
 - Tweak parallelization strategy to use raw threads instead of `Parallel.ForEach`
   - This allows us to easily control thread lifetime, as well as providing a means to eliminate old and no-longer used `CorInfoImpl` instances.
   - This should address to some extent high levels of memory needed for compilation seen by some customers

# Inifite codegen stress mode 
Enable by passing --codegenopt:InfiniteCompileStress=1 on the command line

A simple way to use this would be to run stress it on system.private.corelib.dll.

Assuming that a release clr+libs has been built and your enlistment is in `c:\gitdir\runtime`, run something like the following
```
pushd c:\gitdir\runtime
src\tests\build generatelayoutonly
set CORE_ROOT=C:\gitdir\runtime\artifacts\tests\coreclr\windows.x64.Release\tests\core_root
SET DOTNET_gcServer=1
SET DOTNET_GCHeapCount=6
%CORE_ROOT%\corerun C:\gitdir\runtime\artifacts\bin\coreclr\windows.x64.Release\crossgen2\crossgen2.dll -o:C:\temp\System.Private.CoreLib.dll -r:C:\gitdir\runtime\artifacts\bin\coreclr\windows.x64.Release\IL\*.dll --targetarch:x64 -m:C:\gitdir\runtime\artifacts\bin\coreclr\windows.x64.Release\StandardOptimizationData.mibc --embed-pgo-data -O C:\gitdir\runtime\artifacts\bin\coreclr\windows.x64.Release\IL\System.Private.CoreLib.dll   --pdb --pdb-path:C:\temp --codegenopt:InfiniteCompileStress=1

```

That will run for an infinite period of time, and stress the GC and various other aspects of the runtime.

davidwrighton committed:
Add infinite codegen stress feature to crossgen2, This is known to have a few memory leaks. Enable by passing --codegenopt:InfiniteCompileStress=1 on the command line

jkoritzinsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -744,8 +748,19 @@ void CompileOneMethod(DependencyNodeCore<NodeFactory> dependency)
+                        }
+                        else
+                        {
+                            if ((s_methodsCompiledPerThread % 1000) == 0)
```
Is this condition part of the stress mode or is it something we're going to keep in the product?

davidwrighton replied:
Something like this is something we should keep in the product. We have had periodic complaints about this growing without bound, and simply cleaning up periodically isn't terrible. Every 1000 methods is probably a bit too much/not enough, and it really should be tied to the size of the data structures being held onto though.

jkoritzinsky replied:
If we keep this in the product, can we add a comment here about why we're doing it?

davidwrighton committed:
Improve memory behavior of crossgen2
- Remove memory leak involving UnboxingMethodDesc/PInvokeTargetNativeMethod where they were stored into long term data structures instead of only being used within the confines of the JIT
 - Tweak IsEffectivelySealed to work off of the unboxed method
 - Only call IsEffectivelySealed on Virtual methods
 - Call VersionsWithMethodBody after converting PInvokeTargetNativeMethod to its underlying method
- Our use of ConcurrentDictionary was not memory friendly. In particular, it would typically allocate a new delegate to check to see if an object was in the dictionary in our various Concurrent operations. This is fixed by explicitly controlling delegate lifetime.

davidwrighton committed:
Reduce transient memory allocations
- re-implement UnderlyingType calculation to avoid allocation of GetFields temps
- Adjust implementation of IsClassPreInited to use a cache of already computed data, and optimize use of GetFields
- Cache a MemoryStream on the CorInfoImpl that will be used to build PGO data array. This avoids reallocating it excessively
- Adjust use of TokenResolver to avoid needless allocations of additional TokenResolverProvider through the rental strategy and some careful extraction of code from System.Reflection.Metadata

davidwrighton committed:
Use custom parallelizing approach
- Allows explicit control of thread local state, amount of parallelization and for cleanup of CorInfoImpl instances before beginning the memory intensive layout process

davidwrighton committed:
Final tweaks

davidwrighton committed:
Remove PInvokeTargetNative changes as they break NativeAOT which has different rules for lifetime

trylek commented on src/coreclr/tools/Common/JitInterface/UnboxingMethodDesc.cs:
```diff
@@ -74,6 +76,13 @@ protected override int CompareToImpl(MethodDesc other, TypeSystemComparer compar
+
+        protected override int ComputeHashCode()
+        {
+            throw new NotSupportedException("This method may not be stored  as it is expected to only be used transiently in the JIT");
```
Nice, thanks for adding the check modulo the nit regarding double blank before 'as' ;-).

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs:
```diff
@@ -229,12 +317,35 @@ private class TokenResolverProvider : ISignatureTypeProvider<DummyTypeInfo, Modu
                 _contextModule = contextModule;
             }
 
+            public static TokenResolverProvider Rent(ModuleTokenResolver resolver, IEcmaModule contextModule)
```
Just curious - wouldn't it be even more performant to use a <code>ConcurrentDictionary&lt;IEcmaModule, TokenResolverProvider&gt;</code> provided we flush it after each compilation? While the <code>TokenResolverProvider</code> rental is a smart micro-optimization, it doesn't prevent repeated provider creation for the same context module which seems superfluous - does it mean that the occasional repeated allocation is still cheaper than using a dictionary for unification?

davidwrighton replied:
Eh, this is probably cheaper, as we almost never actually need to allocate a new one, and TLS lookup is really quite fast.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -543,7 +553,7 @@ private static bool FunctionJustThrows(MethodIL ilBody)
         {
             // If any il bodies need to be recomputed, force recompilation
-            if (_ilBodiesNeeded != null)
+            if ((_ilBodiesNeeded != null) || InfiniteCompileStress.Enabled)
```
Does this mean that with <code>InfiniteCompileStress</code> Crossgen2 literally never finishes? Would it be more general to use some such variable to define the number of compilation iterations so that we could subsequently create a pipeline with the option set to something like 100 (that might even roughly mimic the GC stress slowdown)?

davidwrighton replied:
That sounds interesting. But of course, that wouldn't be "infinite" stress. We're already in a state where seeing a real problem takes days.

trylek reviewed (approved):
Looks great to me, thank you!

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DevirtualizationManager.cs:
```diff
@@ -22,6 +23,10 @@ public override bool IsEffectivelySealed(TypeDesc type)
 
         public override bool IsEffectivelySealed(MethodDesc method)
         {
+            if (method.IsUnboxingThunk())
+            {
+                return IsEffectivelySealed(method.GetUnboxedMethod());
+            }
```
I'd do this in JitInterface - unboxing thunks are supposed to be local to JitInterface - they should never escape it (having to add `using Internal.JitInterface` means they escaped).

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:
```diff
@@ -429,6 +429,24 @@ public override IEnumerable<FieldDesc> GetFields()
             }
         }
 
+        public override TypeDesc UnderlyingType
```
Maybe delete the TODO here:

https://github.com/dotnet/runtime/blob/4ff5a3a85a9b0de7ab9e9267959e668142815f4f/src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs#L475

If we're doing overrides, we're unlikely to benefit from a cache there.

davidwrighton replied:
Good point. I've removed the comment.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/Stubs/PInvokeTargetNativeMethod.cs:
```diff
@@ -1,6 +1,8 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
```
Is this needed?

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:
```diff
@@ -81,7 +81,7 @@ public abstract class Compilation : ICompilation
         {
             // Flush the cache when it grows too big
-            if (_methodILCache.Count > 1000)
+            if (_methodILCache.Count > 3000) // This is a magic number found by compiling System.Private.CoreLib.dll that seems to be reasonable.
```
This file is specific to NativeAOT? Did you intend to change this one? ("found by compiling System.Private.CoreLib.dll" doesn't sound like NativeAOT)

davidwrighton replied:
Woops, I wrote that comment and updated the code with the intention of testing it on System.Private.CoreLib, and it seemed to fix a slowdown I was seeing, but clearly it did nothing.    :) I'll revert this one

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/CodeGen/IJitHashableOnly.cs:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    // Mark a type system object with this interface to indicate that it
```
I'm not sure I understand how this avoids putting the objects into long lived storage - we call a different GetHashCode - how does that avert long lived storage?

davidwrighton replied:
Ah, the trick is to make the normal GetHashCode fail with an exception so that special logic needs to be written to make it be stuffable into a collection, but allow a separate path for the hashcode for situations where the JitInterface needs to store the value into a hashtable

MichalStrehovsky replied:
Ah, so this is just a way to guard these instances don't leak out of CorInfoImpl and don't get stored in one of the many collections we have.

Should IJistHashtableOnly go to src\coreclr\tools\Common\JitInterface (and the respective namespace)? It looks odd to include this from the type system project if the purpose is specifically not to make this a general type system entity.

davidwrighton committed:
Fix a couple of bugs and notes from code review

davidwrighton committed:
Respond to code review feedback
